packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  AzureDevOpsCredential,
+  isAzureClientSecretCredential,
+  isAzureManagedIdentityCredential,
+  isAzureDevOpsPersonalAccessToken,
+} from './config';
+import {
+  ClientSecretCredential,
+  ManagedIdentityCredential,
+  TokenCredential,
+} from '@azure/identity';
+import {
+  AzureDevOpsCredentials,
+  AzureDevOpsCredentialsProvider,
+} from './types';
+import { DateTime } from 'luxon';
+
+type CachedAzureDevOpsCredentials = AzureDevOpsCredentials & {
+  expiresAt?: DateTime;
+};
+
+/**
+ * A credentials provider that caches the credentials for as long as it is valid.
+ *
+ * @public
+ */
+export class CachedAzureDevOpsCredentialsProvider
+  implements AzureDevOpsCredentialsProvider
+{
+  azureDevOpsScope = '499b84ac-1321-427f-aa17-267ca6975798/.default';
+  cached: CachedAzureDevOpsCredentials | undefined;
+
+  static create(
+    config: AzureDevOpsCredential,
+  ): CachedAzureDevOpsCredentialsProvider {
+    if (isAzureDevOpsPersonalAccessToken(config)) {
+      return new CachedAzureDevOpsCredentialsProvider(config.token);
+    }
+
+    if (isAzureClientSecretCredential(config)) {
+      return new CachedAzureDevOpsCredentialsProvider(
+        new ClientSecretCredential(
+          config.tenantId,
+          config.clientId,
+          config.clientSecret,
+        ),
+      );
+    }
+
+    if (isAzureManagedIdentityCredential(config)) {
+      return new CachedAzureDevOpsCredentialsProvider(
+        new ManagedIdentityCredential(config.clientId),
+      );
+    }
+
+    throw new Error('Unsupported credential type');
+  }
+
+  private constructor(private readonly credential: TokenCredential | string) {}
+
+  async getCredentials(): Promise<AzureDevOpsCredentials> {
+    if (
+      this.cached === undefined ||
+      (this.cached.expiresAt !== undefined &&
+        DateTime.local() > this.cached.expiresAt)
+    ) {
+      if (typeof this.credential === 'string') {
+        const buffer = Buffer.from(`:${this.credential}`, 'utf8');
+
+        this.cached = {
+          headers: {
+            Authorization: `Basic ${buffer.toString('base64')}`,
+          },
+          type: 'pat',
+          token: this.credential,
+        };
+      } else {
+        const accessToken = await this.credential.getToken(
+          this.azureDevOpsScope,
+        );
+
+        if (accessToken === null) {
```

Comment:
Feeling that this could be `!accessToken` just in case, explicit `null` check feels a bit strict here and a bit tricky to debug if empty return is switched to `undefined`.

Reply:
Fixed 👍